### PR TITLE
feat(added hourly api call; enhanced bid logic)

### PR DESCRIPTION
**Pull request by tomb-cyber**
## Description
- Added the hourly external API call
- Added new functionality to BidController
- Unit tests for BidController

## Changes
- HandleBid now takes the current stock price as a second parameter, so that the price can be passed to it without it being that controller's responsibility. Once the highest level 'system controller' is implemented we can think more in depth, where the price should be kept track of, and if this change is any good.

## Notes

### CODE REVIEW

File: AppleStockAPI.Test/BidControllerTests.cs [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+            public void TestValidBid((double CurrentPrice, Bid TestBid, Response ExpectedRes, int ExpectedCountOfBids) testData)
+            {
+                TestHandleBid(testData);
+            }
+
+            private static IEnumerable<(double, Bid, Response, int)> ValidBidTestCases()
+            {
+                yield return (
+                    100,
+                    new Bid { Price = 100, Quantity = 100 },
```
Comment (PaavoNykanen) [inline AppleStockAPI.Test/BidControllerTests.cs:49]: Do we really need this many tests of the same type? Feels like these are just testing the same thing with the same result with very similar inputs.

File: AppleStockAPI.Test/BidControllerTests.cs [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+            }
+
+        }
+
+
+        [TestFixture]
+        public class AddingValidAndInvalidBidsTestFixture : BidControllerTests
+        {
+            [TestCaseSource(nameof(ValidAndInvalidBidTestCases))]
+            public void TestValidOrInvalidBid((double CurrentPrice, Bid TestBid, Response ExpectedRes, int ExpectedCountOfBids) testData)
```
Comment (PaavoNykanen) [inline AppleStockAPI.Test/BidControllerTests.cs:156]: Why are we testing valid and invalid bids here if valid bids were already tested above? Shouldn't this just be the invalid ones?

File: AppleStockAPI.Test/BidControllerTests.cs [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+                Bid? foundBidWithStringId = controller.GetBidWithId(id.ToString());
+                Assert.Multiple(() => {
+                    Assert.That(foundBidWithGuid, Is.EqualTo(foundBidWithStringId));
+                    Assert.That(controller.GetBids(), Has.Count.EqualTo(3));
+                    Assert.That(controller.GetBidWithId("FakeId"), Is.EqualTo(null));
+                });
+            }
+
+            [TestCase]
+            public void RemovingBidsTest() {
```
Comment (PaavoNykanen) [inline AppleStockAPI.Test/BidControllerTests.cs:267]: I think here there are too many tests inside one test case. This test case is testing for invalid and valid input which I think should be separated. Could also see if you could split the tests so that the ID is given as a test case input and the result is asserted for like [here ](https://github.com/appl-stock-project-org/appl-stock-project/blob/932576a99f62db08de13547133d6435745ccfcca/AppleStockAPI.Test/TradeControllerTests.cs#L27)

File: AppleStockAPI/Controllers/BidController.cs [hunk 4]
```diff
@@ -31,10 +42,50 @@ public class BidController {
...
+		public Bid? GetBidWithId(string id) {
+			return bids.Find(bid => bid.Id.ToString() == id);
+		}
+
+		/// <summary>
+		/// Returns a bid that matches given Guid id
+		/// </summary>
+		/// <param name="id">Bid id</param>
+		/// <returns>A bid, if bids has a bid with given Guid id</returns>
+		public Bid? GetBidWithId(Guid id) {
```
Comment (PaavoNykanen) [inline AppleStockAPI/Controllers/BidController.cs:69]: Do we really need both of these? Shouldn't the string input one be enough?

File: AppleStockAPI.Test/BidControllerTests.cs (hunk 1 above)
Comment (tomb-cyber) [reply AppleStockAPI.Test/BidControllerTests.cs:49]: There's edge cases testing the +/-10%, values that arent on the edges, testing that changing the given stock price has effect, using int and double values, testing that the doubles get rounded correctly. What cases do you think are redundant specifically?

File: AppleStockAPI.Test/BidControllerTests.cs (hunk 2 above)
Comment (tomb-cyber) [reply AppleStockAPI.Test/BidControllerTests.cs:156]: Having valid and invalid bids allows us to check that the bids list is added or not added to correctly. Then again could these be merged to the above one, so that list of test cases is even longer? Possibly.

File: AppleStockAPI.Test/BidControllerTests.cs (hunk 3 above)
Comment (tomb-cyber) [reply AppleStockAPI.Test/BidControllerTests.cs:267]: Yeah thats valid, i was lazy and did just one removing test.

File: AppleStockAPI/Controllers/BidController.cs (hunk 4 above)
Comment (tomb-cyber) [reply AppleStockAPI/Controllers/BidController.cs:69]: We dont need them, but then we can just give the id property without always adding ToString. Then again we dont need the string one at all with the scope of this project, because there wont be ids coming from external sources as strings. I will delete the string one, unless you can see a need for that.

Comment (PaavoNykanen) [reply AppleStockAPI/Controllers/BidController.cs:69]: Yeah deleting either one is fine

File: AppleStockAPI.Test/BidControllerTests.cs (hunk 2 above)
Comment (PaavoNykanen) [reply AppleStockAPI.Test/BidControllerTests.cs:156]: I think I would split the test cases to invalid and valid tests, and in both you can always check that the bid was added with the length of the bids array. So kind of like these [tests](https://github.com/appl-stock-project-org/appl-stock-project/blob/932576a99f62db08de13547133d6435745ccfcca/AppleStockAPI.Test/TradeControllerTests.cs#L50) are done.

Comment (tomb-cyber) [reply AppleStockAPI.Test/BidControllerTests.cs:156]: Then we are not able to test what happens to the list when valid and invalid bids are added alternately. If they are separate, when testing only invalid bids in a row the count would always be zero. With such tests we wouldn't catch for example a breaking change where bids list is always initialized to a whole new list object. That's why I think it's a good idea to test what happens when there's valid and invalid bids.

File: AppleStockAPI.Test/BidControllerTests.cs (hunk 1 above)
Comment (PaavoNykanen) [reply AppleStockAPI.Test/BidControllerTests.cs:49]: I guess they are fine, just feels like a lot of repetition if you test all of those same things multiple times. Maybe splitting these by testing cases could also be beneficial? So you could have one for edge values, one for chaning the current price and one for rounding numbers? They are also kinda long with the current syntax which makes them harder to maintain I think. I'm not familiar with c# testing so I don't know the idea behind the `TestFixture` for example but yeah.

File: AppleStockAPI.Test/BidControllerTests.cs (hunk 2 above)
Comment (PaavoNykanen) [reply AppleStockAPI.Test/BidControllerTests.cs:156]: Thats fair, but in that case I think it might be better to write them the way that you did the `BidsListCountGoesUpTest` test. Much more compact and easier to read and the naming could be better to explain what its testing. And if you're testing for if the bids are added or not added to the bid list I think it's not needed to also test the result and error message etc at the same time. Those should probably be in their own test cases right?

Comment (tomb-cyber) [reply AppleStockAPI.Test/BidControllerTests.cs:156]: Yeah the naming is all over, i should also add some comments for them. And you are right, the responses are unnecessary to test in those cases. I just wanted to reuse the same function because they are quite similar. I could modify the test function so that if a response object is not passed, it wont be tested. Also regarding the syntax, i have never done c# test so i have no idea whats the best way. Thats also why these are a bit half assed because i was so tired of figuring out how this library works

Comment (PaavoNykanen) [reply AppleStockAPI.Test/BidControllerTests.cs:156]: Yeah no problem it's not like this actually matters in such a small project course work :D I'm just thinking that for maintainability and if someone else needs to add more tests to the same controller, it should be as clear as possible what the current tests are testing and how they are testing so that developers don't need to spend too much time learning those things.

Comment (tomb-cyber) [reply AppleStockAPI.Test/BidControllerTests.cs:156]: Yeah agreed. I'll clean up this pr this weekend. You can mark it not approved, so that no one else merges it by accident before

File: AppleStockAPI/Controllers/BidController.cs [hunk 5]
```diff
@@ -1,26 +1,37 @@
...
+		/// <param name="bid">Bid to handle</param>
+		/// <param name="currentStockPrice">Current stock price that determines the range of acceptable prices</param>
         /// </summary>
-		public static Response handleBid(Bid bid, List<Bid> listOfBids) {
+		public Response HandleBid(Bid bid, double currentStockPrice) {
 
-			Response response = new Response();
+			Response response = new();
 
 			bid.Price = Math.Round(bid.Price, 2);
```
Comment (PaavoNykanen) [inline AppleStockAPI/Controllers/BidController.cs:24]: Should we change from `Math.Round()` to `Math.Truncate`? Ref: https://github.com/appl-stock-project-org/appl-stock-project/pull/8#discussion_r1507568010

**Review by PaavoNykanen — APPROVED** [review]
